src/probnum/kernels/_exponentiated_quadratic.py

```diff
@@ -0,0 +1,73 @@
+"""Exponentiated quadratic kernel."""
+
+from typing import Optional
+
+import numpy as np
+import scipy.spatial.distance
+
+import probnum.utils as _utils
+from probnum.type import IntArgType, ScalarArgType
+
+from ._kernel import Kernel
+
+_InputType = np.ndarray
+
+
+class ExpQuad(Kernel[_InputType]):
+    """Exponentiated quadratic / RBF kernel.
+
+    Covariance function defined by :math:`k(x_0, x_1) = \\exp \\big(-\\frac{\\lVert
+    x_0 - x_1 \\rVert^2}{2l^2}\\big)`. This kernel is also known as the squared
+    exponential or radial basis function kernel.
+
+    Parameters
+    ----------
+    input_dim :
+        Input dimension of the kernel.
+    lengthscale
+        Lengthscale of the kernel. Describes the input scale on which the process
+        varies.
+
+    See Also
+    --------
+    RatQuad : Rational quadratic kernel.
+    Matern : Matern kernel.
+
+    Examples
+    --------
+    >>> import numpy as np
+    >>> from probnum.kernels import ExpQuad
+    >>> K = ExpQuad(input_dim=1, lengthscale=0.1)
+    >>> K(np.linspace(0, 1, 3)[:, None])
+    array([[1.00000000e+00, 3.72665317e-06, 1.92874985e-22],
+           [3.72665317e-06, 1.00000000e+00, 3.72665317e-06],
+           [1.92874985e-22, 3.72665317e-06, 1.00000000e+00]])
+    """
+
+    def __init__(self, input_dim: IntArgType, lengthscale: ScalarArgType = 1.0):
+        self.lengthscale = _utils.as_numpy_scalar(lengthscale)
+        super().__init__(input_dim=input_dim, output_dim=1)
+
+    def __call__(self, x0: _InputType, x1: Optional[_InputType] = None) -> np.ndarray:
+        # Check and reshape inputs
+        x0, x1, equal_inputs = self._check_and_transform_input(x0, x1)
+        x0_originalshape = x0.shape
+        x1_originalshape = x1.shape
+
+        # Compute pairwise euclidean distances ||x0 - x1|| / l
+        x0 = np.atleast_2d(x0)
+        if equal_inputs:
+            pdists = scipy.spatial.distance.squareform(
+                scipy.spatial.distance.pdist(x0 / self.lengthscale, metric="euclidean")
+            )
+        else:
+            x1 = np.atleast_2d(x1)
+            pdists = scipy.spatial.distance.cdist(
+                x0 / self.lengthscale, x1 / self.lengthscale, metric="euclidean"
+            )
+
+        # Kernel matrix
+        kernmat = np.exp(-(pdists ** 2) / 2.0)
+        return self._transform_kernelmatrix(
+            kerneval=kernmat, x0_shape=x0_originalshape, x1_shape=x1_originalshape
+        )
```

Comment:
So do I understand correctly that the decision whether `__call__` returns a kernel matrix or a single kernel evaluation is made by specifying  `input_dim` and then checking at the call, whether the user put in a single point or a stack of points? That is a good idea, and certainly one way of doing this.

Though, and this is just to mention a possible alternative, not because I think it is superior, wouldnt the interface be more didactic if there was a function `__call__(self, ...)` and another one `matrix(self, ...)` both of which would do different things? This would have the advantage that if something of the wrong shape is passed, an error is raised, instead of doing something that the user does not expect. In this case, one could also avoid the `input_dim` decision at initialisation. This might be irrelevant, but in my code I at some point converged to the `call` *and* the `matrix`, because (i) it was easier to test, and (ii) saved *a lot* of case distinctions in each function.

I am not saying this is better, but I'd like to hear your opinion on this decision :) Thanks a lot for all of this work btw!

Reply:
I've thought about this a bit and I have decided against separating them for the following reasons.
1. The user can still pass only two vectors to `matrix` and will obtain the same behaviour as for `__call__` which means there are two ways of doing the same thing without a clear separation of which one is the correct one. In particular one returns a single entry matrix and another a scalar. This breaks the way we specify covariances to one dimensional random variables.
2. While one would save the argument `input_dim` and `_check_and_transform_input` one would have to specify two functions for each subclass of `Kernel`, namely `__call__` and `matrix` which would lead to some code duplication. 

That being said, one cannot evaluate the diagonal of the kernel matrix without forming the full kernel matrix in the current setup. One solution would be your suggestion with the above caveats, or adding an argument to `__call__(x0, x1, diag=False)`, which can then offer a time/space efficient function to compute just the diagonal of the kernel matrix.